core/src/main/java/org/apache/iceberg/actions/SizeBasedFileRewriter.java

```diff
@@ -0,0 +1,317 @@
+   */
+  protected long numOutputFiles(long inputSize) {
+    if (inputSize < targetFileSize) {
+      return 1;
+    }
+
+    long numFilesWithRemainder = LongMath.divide(inputSize, targetFileSize, RoundingMode.CEILING);
+    long numFilesWithoutRemainder = LongMath.divide(inputSize, targetFileSize, RoundingMode.FLOOR);
+    long avgFileSizeWithoutRemainder = inputSize / numFilesWithoutRemainder;
+
+    if (LongMath.mod(inputSize, targetFileSize) > minFileSize) {
+      // the remainder file is of a valid size for this rewrite so keep it
+      return numFilesWithRemainder;
+
+    } else if (avgFileSizeWithoutRemainder < Math.min(1.1 * targetFileSize, writeMaxFileSize())) {
+      // if the reminder is distributed amongst other files,
+      // the average file size will be no more than 10% bigger than the target file size
+      // so round down and distribute remainder amongst other files
+      return numFilesWithoutRemainder;
+
+    } else {
+      // keep the remainder file as it is not OK to distribute it amongst other files
+      return numFilesWithRemainder;
+    }
+  }
+
+  /**
+   * Estimates a larger max target file size than the target size used in task creation to avoid
+   * tasks which are predicted to have a certain size, but exceed that target size when serde is
+   * complete creating tiny remainder files.
+   *
+   * <p>While we create tasks that should all be smaller than our target size, there is a chance
+   * that the actual data will end up being larger than our target size due to various factors of
+   * compression, serialization and other factors outside our control. If this occurs, instead of
+   * making a single file that is close in size to our target, we would end up producing one file of
+   * the target size, and then a small extra file with the remaining data. For example, if our
+   * target is 512 MB, we may generate a rewrite task that should be 500 MB. When we write the data
+   * we may find we actually have to write out 530 MB. If we use the target size while writing we
+   * would produce a 512 MB file and an 18 MB file. If instead we use a larger size estimated by
+   * this method, then we end up writing a single file.
+   *
+   * @return the target size plus one half of the distance between max and target
+   */
+  protected long writeMaxFileSize() {
+    return (long) (targetFileSize + ((maxFileSize - targetFileSize) * 0.5));
+  }
+
+  private Map<String, Long> sizeThresholds(Map<String, String> options) {
+    long target = propertyAsLong(options, TARGET_FILE_SIZE_BYTES, defaultTargetFileSize());
+
+    long defaultMin = (long) (target * MIN_FILE_SIZE_DEFAULT_RATIO);
+    long min = propertyAsLong(options, MIN_FILE_SIZE_BYTES, defaultMin);
+
+    long defaultMax = (long) (target * MAX_FILE_SIZE_DEFAULT_RATIO);
+    long max = propertyAsLong(options, MAX_FILE_SIZE_BYTES, defaultMax);
+
+    checkArgument(target > 0, "'%s' is set to %s but must be > 0", TARGET_FILE_SIZE_BYTES, target);
```

Comment:
I switched negation in error messages. I believe it is easier to understand an error message that states what the value should look like, instead of what it can’t be.